### PR TITLE
[Feature] Improve and fix Prometheus & Grafana integrations

### LINKED ISSUE #871: [Feature]  Enable Grafana Dashboard integration with KubeRay
### Search before asking

- [X] I had searched in the [issues](https://github.com/ray-project/kuberay/issues) and found no similar feature requirement.

### Description

Currently, Ray exposes all Grafana files on the head node in the `session_latest` directory. See [docs](https://docs.ray.io/en/latest/ray-observability/ray-metrics.html#grafana).

Enable KubeRay the ability to configure a Grafana Instance and Prometheus instance (which is NOT running in the same container as the Ray Head Node). 

Integrate Ray Dashboard UI to display graphs from Grafana.

Potential solutions: add configuration options through the ray-cluster yaml?

### Use case

Allow the user to integrate external Grafana and Prometheus instances with Ray while utilizing the default dashboards.

### Related issues

NA

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by kevin85421**
## Why are these changes needed?

The old document (observability.md) of Prometheus and Grafana has some issues that make users hard to follow.

* Inconsistency between **install.sh** and **observability.md**
  * ServiceMonitor example in the doc and ServiceMonitor instance created by the script. 
  * PodMonitor example in the doc and PodMonitor instance created by the script. 

* Inconsistency between **observability.md** and KubeRay operator [kuberay/#230](https://github.com/ray-project/kuberay/pull/230).
  *  The document uses port 9001 as a metrics endpoint, but KubeRay operator uses 8080 as a default metrics endpoint.

* Some missing pieces:
  * Both examples of ServiceMonitor and PodMonitor in observability.md lack some labels to match `serviceMonitorSelector` and `podMonitorSelector`.
  * How to access Prometheus UI?
  * How to access Grafana UI?
  * How to import Ray Dashboard into Grafana?

## Related issue number

Closes #871 

## Checks

- [ ] I've made sure the tests are passing. 
- Testing Strategy
   - [ ] Unit tests
   - [x] Manual tests
   - [ ] This PR is not tested :(

prometheus_web_ui

* Prometheus UI: Prometheus can detect both ServiceMonitor (ray-head-monitor) and PodMonitor (ray-worker-monitor).

grafana_ray_dashboard

* Grafana UI: Load Ray Dashboard into Grafana via `config/grafana/dashboard_default.json`.

### CODE REVIEW

**Comment by kevin85421** [comment]
cc @Yicheng-Lu-llll would you mind reviewing this PR? Thanks!

File: docs/guidance/prometheus-grafana.md [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+# deployment.apps/prometheus-grafana                    1/1     1            1           46s
+# deployment.apps/prometheus-kube-prometheus-operator   1/1     1            1           46s
+# deployment.apps/prometheus-kube-state-metrics         1/1     1            1           46s
+```
+* KubeRay provides an [install.sh script](../../install/prometheus/install.sh) to install the [kube-prometheus-stack](https://github.com/prometheus-community/helm-charts/tree/main/charts/kube-prometheus-stack) chart and related custom resources, including **ServiceMonitor** and **PodMonitor**, in the namespace `prometheus-system` automatically.
+
+## Step 3: Install a KubeRay operator
+
+```sh
+helm repo add kuberay https://ray-project.github.io/kuberay-helm/
```
Comment (gvspraveen) [inline docs/guidance/prometheus-grafana.md:33]: instead of duplicating this and potentially becoming out of date, should we just point to `helm-chart/kuberay-operator/README.md`?

same with ray cluster below

Comment (kevin85421) [reply docs/guidance/prometheus-grafana.md]: Good point. Updated https://github.com/ray-project/kuberay/pull/895/commits/4bc3a748c31981cbc0659c53614546f4758e588a.

File: docs/guidance/prometheus-grafana.md [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+  # NAME                                          READY   STATUS    RESTARTS   AGE
+  # raycluster-kuberay-worker-workergroup-5stpm   1/1     Running   0          3h16m
+  ```
+
+* `ray.io/cluster: $RAY_CLUSTER_NAME`: We also define `metadata.labels` by manually adding `ray.io/cluster: <ray-cluster-name>` and then instructing the PodMonitors resource to add that label in the scraped metrics via `spec.podTargetLabels[0].ray.io/cluster`.
+
+## Step 7: Access Prometheus Web UI
+```sh
+# Forward the port of Prometheus Web UI in the Prometheus server Pod.
+kubectl port-forward --address 0.0.0.0 prometheus-prometheus-kube-prometheus-prometheus-0 9090:9090
```
Comment (Yicheng-Lu-llll) [inline docs/guidance/prometheus-grafana.md:159]: I have run this PR and it all looks good to me except this should be:
```
kubectl port-forward --address 0.0.0.0 prometheus-prometheus-kube-prometheus-prometheus-0 -n prometheus-system 9090:9090
```

Comment (kevin85421) [reply docs/guidance/prometheus-grafana.md]: Good catch! Update and test it again.

**Review by Yicheng-Lu-llll — APPROVED** [review]

**Comment by kevin85421** [comment]
Gentle ping @scarlet25151. Would you mind taking a look at this PR? Thank you!

File: docs/guidance/prometheus-grafana.md [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+* KubeRay provides an [install.sh script](../../install/prometheus/install.sh) to install the [kube-prometheus-stack](https://github.com/prometheus-community/helm-charts/tree/main/charts/kube-prometheus-stack) chart and related custom resources, including **ServiceMonitor** and **PodMonitor**, in the namespace `prometheus-system` automatically.
+
+## Step 3: Install a KubeRay operator
+
+* Follow this [document](../../helm-chart/kuberay-operator/README.md) to install the latest stable KubeRay operator via Helm repository.
+
+## Step 4: Install a RayCluster
+
+```sh
+helm install raycluster kuberay/ray-cluster --version 0.4.0
```
Comment (architkulkarni) [inline docs/guidance/prometheus-grafana.md:37]: @kevin85421 The tutorial failed for me at this step.  Output:

```
helm install raycluster kuberay/ray-cluster --version 0.4.0

Error: INSTALLATION FAILED: unable to build kubernetes objects from release manifest: resource mapping not found for name: "raycluster-kuberay" namespace: "" from "": no matches for kind "RayCluster" in version "ray.io/v1alpha1"
ensure CRDs are installed first
```

File: docs/guidance/prometheus-grafana.md [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+
```
Comment (architkulkarni) [inline docs/guidance/prometheus-grafana.md:38]: Consider explicitly adding a command here for how to get the RAYCLUSTER_HEAD_POD needed in the next step.   I used `kubectl get pods`

Comment (architkulkarni) [reply docs/guidance/prometheus-grafana.md:38]: Maybe also indicate that the user needs to watch this command until the pod is RUNNING (though the error message is pretty easy to understand: `error: unable to forward port because pod is not running. Current status=Pending`)

File: docs/guidance/prometheus-grafana.md [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+# Forward the port of the Prometheus metrics endpoint.
+kubectl port-forward --address 0.0.0.0 ${RAYCLUSTER_HEAD_POD} 8080:8080
+curl localhost:8080
```
Comment (architkulkarni) [inline docs/guidance/prometheus-grafana.md:41]: We should say that the user needs to open a new terminal here (since the `port-forward` command blocks)

File: docs/guidance/prometheus-grafana.md [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+# and "serviceMonitor/prometheus-system/ray-head-monitor/0 (1/1 up)" in the page.
+```
+
+![Prometheus Web UI](../images/prometheus_web_ui.png)
+
+## Step 8: Access Grafana
+
+```sh
+# Forward the port of Grafana
+kubectl port-forward --address 0.0.0.0 deployment/prometheus-grafana -n prometheus-system 3000:3000
```
Comment (architkulkarni) [inline docs/guidance/prometheus-grafana.md:172]: Got this error at this step:
```
Unable to listen on port 3000: Listeners failed to create with the following errors: [unable to create listener: Error listen tcp4 0.0.0.0:3000: bind: address already in use]
error: unable to listen on any of the requested ports: [{3000 3000}]
```

But I'm not sure what could be using port 3000.  Do I need to cancel one of the previous steps?

Comment (architkulkarni) [reply docs/guidance/prometheus-grafana.md:172]: Looks like I already have grafana running for some reason, I think it's unrelated to this tutorial: 
```
sudo lsof -i :3000
Password:
COMMAND    PID   USER   FD   TYPE             DEVICE SIZE/OFF NODE NAME
grafana-s 3915 archit   21u  IPv6 0xd6d807306e5fa78f      0t0  TCP *:hbci (LISTEN)
```

Let me figure out how to stop it. I don't know if it's a common enough situation to mention in this document.

Comment (architkulkarni) [reply docs/guidance/prometheus-grafana.md:172]: Naively killed the process but it kept restarting. The solution here https://unix.stackexchange.com/questions/601063/port-3000-is-always-being-hogged-by-grafana-server
is to run `brew services stop grafana`.  Probably don't need to mention in this doc

**Review by architkulkarni — APPROVED** [review]
Went through the tutorial, it all works!  Left some minor suggestions.

File: docs/guidance/prometheus-grafana.md (hunk 4 above)
Comment (kevin85421) [reply docs/guidance/prometheus-grafana.md:38]: Updated. https://github.com/ray-project/kuberay/pull/895/commits/b990af9fac79ff070e0024429efef4bd0f0141e4

File: docs/guidance/prometheus-grafana.md (hunk 5 above)
Comment (kevin85421) [reply docs/guidance/prometheus-grafana.md:48]: Updated. https://github.com/ray-project/kuberay/pull/895/commits/b990af9fac79ff070e0024429efef4bd0f0141e4
